lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -2238,41 +2225,14 @@ export class UserAgentApplication {
      * Utils function to create the Authentication
      * @param {@link account} account object
      * @param scopes
-     * @param silentCall
      *
-     * @returns {string} token type: id_token or access_token
+     * @returns {string} token type: token, id_token or id_token token
      *
      */
-    private getTokenType(accountObject: Account, scopes: string[], silentCall: boolean): string {
-        /*
-         * if account is passed and matches the account object/or set to getAccount() from cache
-         * if client-id is passed as scope, get id_token else token/id_token_token (in case no session exists)
-         */
-        let tokenType: string;
-
-        // acquireTokenSilent
-        if (silentCall) {
-            if (Account.compareAccounts(accountObject, this.getAccount())) {
-                tokenType = (scopes.indexOf(this.config.auth.clientId) > -1) ? ResponseTypes.id_token : ResponseTypes.token;
-            }
-            else {
-                tokenType  = (scopes.indexOf(this.config.auth.clientId) > -1) ? ResponseTypes.id_token : ResponseTypes.id_token_token;
-            }
-
-            return tokenType;
-        }
-        // all other cases
-        else {
-            if (!Account.compareAccounts(accountObject, this.getAccount())) {
-                tokenType = ResponseTypes.id_token_token;
-            }
-            else {
-                tokenType = (scopes.indexOf(this.clientId) > -1) ? ResponseTypes.id_token : ResponseTypes.token;
-            }
-
-            return tokenType;
-        }
-
+    private getTokenType(accountObject: Account, scopes: string[]): string {
+        const accountsMatch = Account.compareAccounts(accountObject, this.getAccount());
+        const loginScopesOnly = ScopeSet.onlyContainsClientId(scopes, this.clientId) || ScopeSet.onlyContainsOidcScopes(scopes);
```

Comment:
Why not do the clientId translation when validating the request and then you only have to call `onlyContainsOidcScopes` here?

Reply:
Then instead of having the normalize function that translates then appends default scopes you can just append the default scopes. The flow make's a little bit more sense to me.

Reply:
Yeah, that definitely sounds better, implementing.

Reply:
Implementing this change actually revealed a bunch of other places where clientId was being added as a single scope for requests that had null scopes. In all of the found cases, clientId has been replaced with OIDC scopes array to avoid having clientId being added after validation, to avoid clientId ending up in the final scopes array. In other words, since we now translate during validation and not at the end, if the clientId was added post-validation, it wouldn't be caught.

Reply:
Remember, you cannot mutate user passed scopes, and if we are returning what is requested (not sure if we do), we need to retain the user passed scopes info. Having said that, I agree with @tnorling.